Kernel/FileSystem/RAMFS/Inode.cpp

```diff
@@ -149,13 +183,44 @@ ErrorOr<size_t> RAMFSInode::write_bytes_to_content_space(size_t offset, size_t i
     return do_io_on_content_space(*mapping_region, offset, io_size, const_cast<UserOrKernelBuffer&>(buffer), true);
 }
 
-ErrorOr<size_t> RAMFSInode::write_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer const& buffer, OpenFileDescription*)
+ErrorOr<size_t> RAMFSInode::write_bytes_locked_increasing_allocated_blocks_count(size_t last_block_index, off_t offset, size_t size, UserOrKernelBuffer const& buffer, OpenFileDescription* description)
 {
     VERIFY(m_inode_lock.is_locked());
     VERIFY(!is_directory());
     VERIFY(offset >= 0);
 
-    TRY(ensure_allocated_blocks(offset, size));
+    auto original_last_block_index = m_blocks.size();
+    VERIFY(original_last_block_index < last_block_index);
+    TRY(change_blocks_list_size(TruncateToLastBlock::No, last_block_index));
+
+    ArmedScopeGuard clean_allocated_blocks_on_failure([&] {
+        for (size_t block_index = original_last_block_index; block_index < last_block_index; block_index++) {
+            if (m_blocks[block_index])
+                m_blocks[block_index].clear();
+        }
+        MUST(change_blocks_list_size(TruncateToLastBlock::Yes, original_last_block_index));
+    });
+
+    auto nwritten = TRY(write_bytes_locked_with_same_blocks_count(offset, size, buffer, description));
+    clean_allocated_blocks_on_failure.disarm();
+    return nwritten;
+}
+
+ErrorOr<size_t> RAMFSInode::write_bytes_locked_with_same_blocks_count(off_t offset, size_t size, UserOrKernelBuffer const& buffer, OpenFileDescription*)
+{
+    VERIFY(m_inode_lock.is_locked());
+    VERIFY(!is_directory());
+    VERIFY(offset >= 0);
+
+    VERIFY(!Checked<size_t>::addition_would_overflow(offset, size));
+    VERIFY(!Checked<size_t>::addition_would_overflow(((offset + size) / DataBlock::block_size), (((offset + size) % DataBlock::block_size) == 0 ? 0 : 1)));
+    size_t last_block_index = ((offset + size) / DataBlock::block_size) + (((offset + size) % DataBlock::block_size) == 0 ? 0 : 1);
```

Comment:
If we use an actual `Checked<size_t>` instead of the helper we can pull out the result of the multiplication directly while checking that it didn't overflow. That way, we can avoid writing everything twice.